Source/Scene/VoxelPrimitive.js

```diff
@@ -0,0 +1,2514 @@
+  const keyframeCount = that._keyframeCount;
+  const despeckle = that._despeckle;
+  const jitter = that._jitter;
+  const nearestSampling = that._nearestSampling;
+  const customShader = that._customShader;
+  const attributeLength = types.length;
+
+  const shaderBuilder = new ShaderBuilder();
+
+  // Vertex shader
+  shaderBuilder.addVertexLines(["#line 0", VoxelVS]);
+  shaderBuilder.setPositionAttribute("vec4", "a_position");
+
+  // Fragment shader
+  shaderBuilder.addFragmentLines([
+    customShader.fragmentShaderText,
+    "#line 0",
+    VoxelFS,
+  ]);
+
+  const fragmentStructId = "FragmentInput";
+  const fragmentStructName = "FragmentInput";
+  const attributeStructId = "Attributes";
+  const attributeStructName = "Attributes";
+  const attributeFieldName = "attributes";
+  const voxelStructId = "Voxel";
+  const voxelStructName = "Voxel";
+  const voxelFieldName = "voxel";
+
+  // Attribute struct
+  shaderBuilder.addStruct(
+    attributeStructId,
+    attributeStructName,
+    ShaderDestination.FRAGMENT
+  );
+
+  for (let i = 0; i < attributeLength; i++) {
+    const name = names[i];
+    const type = types[i];
+    const glslType = getGlslType(type);
+    shaderBuilder.addStructField(attributeStructId, glslType, name);
+  }
+
+  // Voxel struct
+  shaderBuilder.addStruct(
+    voxelStructId,
+    voxelStructName,
+    ShaderDestination.FRAGMENT
+  );
+
+  for (let i = 0; i < attributeLength; i++) {
+    const name = names[i];
+    const type = types[i];
+    const glslType = getGlslType(type);
+    shaderBuilder.addStructField(voxelStructId, glslType, `${name}Minimum`);
+    shaderBuilder.addStructField(voxelStructId, glslType, `${name}Maximum`);
+    shaderBuilder.addStructField(voxelStructId, "vec3", `${name}NormalLocal`);
+    shaderBuilder.addStructField(voxelStructId, "vec3", `${name}NormalWorld`);
+    shaderBuilder.addStructField(voxelStructId, "vec3", `${name}NormalView`);
+    shaderBuilder.addStructField(voxelStructId, "bool", `${name}NormalValid`);
+  }
+
+  shaderBuilder.addStructField(voxelStructId, "vec3", "positionEC");
+  shaderBuilder.addStructField(voxelStructId, "vec3", "positionUv");
+  shaderBuilder.addStructField(voxelStructId, "vec3", "positionUvShapeSpace");
+  shaderBuilder.addStructField(voxelStructId, "vec3", "positionUvLocal");
+  shaderBuilder.addStructField(voxelStructId, "vec3", "viewDirUv");
+  shaderBuilder.addStructField(voxelStructId, "vec3", "viewDirWorld");
+  shaderBuilder.addStructField(voxelStructId, "float", "travelDistance");
+
+  // FragmentInput struct
+  shaderBuilder.addStruct(
+    fragmentStructId,
+    fragmentStructName,
+    ShaderDestination.FRAGMENT
+  );
+  shaderBuilder.addStructField(
+    fragmentStructId,
+    attributeStructName,
+    attributeFieldName
+  );
+  shaderBuilder.addStructField(
+    fragmentStructId,
+    voxelStructName,
+    voxelFieldName
+  );
+
+  shaderBuilder.addUniform(
+    "sampler2D",
+    "u_megatextureTextures[METADATA_COUNT]",
```

Comment:
Maybe I haven't come across it yet in this PR, but it would be good to document what exactly you mean by "megatexture" somewhere, I don't know the memory layout or type

Reply:
Megatexture is handled in VoxelTraversal but I don't think it's documented very well. In short, it's a large texture that stores multiple tiles of voxel data and pages tiles in and out as the camera moves.